### ISSUE TITLE
make it easy to follow a link

### ISSUE THREAD

**Issue by aisbergde** [marktext/marktext#2165]
### Describe your feature request

I would like to click on a link to **easy** follow the link

### What problem does this feature solve?

Currently it is hard or not possible to follow a link. Sometimes (I don't understand the logic behind) get a small context menu containing 2 buttons where I could click the second button to follow the link. But this works only sometimes in preview mode, but not in raw edit mode at all. 

Sometimes I get this 2 button menu, but I need to hurry up to use it because it hides very quickly (after half second maybe).

There is also a "normal" context menu (I am using Windows), but there is also no way to follow the link. 

Sometimes I don't want to use the special Markdown syntax but only write http://aaa.bbb and this should also be clickable.

I was never able to follow a link from inside marktext. Finally I have to copy and past the link into a browser. terrible. This is a really show stopper. And finally I have to use VS Code or any other markdown editor where I can follow links.

**Comment by aisbergde** [comment]
Here is an idea how typora does, there is a "open Link" in the context menu

![image](https://user-images.githubusercontent.com/9020187/83618071-3dda2d80-a58a-11ea-974e-a483229c815c.png)

**Comment by fxha** [comment]
> Currently it is hard or not possible to follow a link. Sometimes (I don't understand the logic behind) get a small context menu containing 2 buttons where I could click the second button to follow the link. But this works only sometimes in preview mode, but not in raw edit mode at all.

You need to hover over the link to get the overlay or you could also use `Ctrl+Click` to open the link directly. I think a context menu is not nedded but we should definitely create a tutorial soon or help pages for this small details.

**Comment by aisbergde** [comment]
to hoover and to get this small menu is very hard to do as I described. Most time it does not work. 

I will try to ctrl-click. But how a user should know about? It did not come in my mind. That's why an entry in the  context menu with a hint "ctrl-click" is like an inline documentation. This would be intuitive and no need to search the documentation when someone want to follow the link. of course it should be *also* in the documentation.

**Comment by fxha** [comment]
> I will try to ctrl-click. But how a user should know about? It did not come in my mind. 

It's a default shortcut in almost all editors/office suite applications but I agree that we should add this to the documentation for better UX.

**Comment by aisbergde** [comment]
of course you are right. But I use different programs with different shortcuts and instead of thinking about which shortcut is supported in the current program I normally look at the shortcut hints in the menu and context menu and if there is no hint I even don't try. That's why I suggest to include at least the most important actions in the menu - as a reminder: "it is there and it works as in nearly all editors".

**Comment by proxseas** [comment]
Ctrl-click'ing doesn't work for me. This is really buggy. The only way to open a link is to manually copy & paste it for me right now.

**Comment by brainchild0** [comment]
It seems that if a URI appears directly in the text, then the application treats it as a link, by changing the cursor and handing a *ctrl*-*click* to visit the link target, but does not display a popup item. If a link is formatted with text, using Markdown syntax, then the application also displays a popup of two icons. One icon is for visiting the link target, while the other is for removing the text from the link.

Overall, this behavior is confusing. 

The *ctrl*-*click* handling is standard, and should be preserved. A popup icon for visiting the link target may be helpful in some cases, as long as it is not intrusive. If the interface is intuitive, then it will be usable without a tutorial. 

Editing features such as removing the link text, however, should be presented in the same way as other editing features, for example, changing letter style. And behavior should be as much the same as reasonable for true links versus simple URIs.

**Comment by brainchild0** [comment]
From observation, it appears that factors that currently prevent the popup icons from appearing, in addition to the link being represented as a simple URI (see above), are formatting inside the link text (e.g. `[*text*](http://some.where/)`), and the editing caret  being positioned inside the link at the same time.

Further, links represented with malformed syntax (e.g. `[foo]((http://some/bar)`, _notice extra open parenthesis_) are rendered as links and may cause the popup icons, yet the action of visiting the link target is ignored.

All of these behaviors are confusing. The actual behavior should be simplified into  a few different cases, each appropriate for the situation.

**Comment by dzmitryv** [comment]
I've just tried `ctrl+click`. It doesn't work for me. I'm running MarkText v0.17.1 on Linux. 
I tried both types of link: 
- markdown link `[foo](http://some/where)`
- bare URI `http://some/where`

I'm mentioning this because I haven't seen Linux mentioned in the comments above. Maybe that's an issue with running on that OS.

**Comment by fxha** [comment]
@dzmitryv You need a valid URL like `https://github.com` or local path like `/tmp/abc.pdf`, otherwise the source will not be opened for safety reasons.

**Comment by dzmitryv** [comment]
@fxha sorry, I don't understand what exactly does the URLs that I used as an example not valid. I think they as valid as `https://github.com` is. Could you please elaborate?

**Comment by fxha** [comment]
The last one `http://foo.bar.baz` works, but I expect an issue with our URL validation.

**Comment by dzmitryv** [comment]
@brainchild0 , I've just re-tested the following:

```
1. inline link [foo](http://go/test) 
2. bare URI http://go/test  
3. autolink <http://go/test>
4. [gitHub](https://github.com)
5. https://github.com
6. <https://github.com>
7. [gitHub](http://github.com)
8. http://github.com
9. <http://github.com>
```

All those URLs are perfectly working in my environment. I mean, they all lead to some web-site if I use them in browser, or if I open the same .md file for instance in Atom editor with Markdow Preview Enhanced package and click on those links.

What I got in MarkText:
- **Pop-up icons appear** for 1, 4 and 7 when I hover mouse pointer over them. The icons don't appear for the other items.
- **Click on the link icon opens the URL in browser** for 4 and 7. Nothing happens for 1.
- **Ctrl+click works** for 4-9.
- Item 2 (bare URI) **doesn't get styled as link** (it has the same color as plain text and isn't underscored).

**Comment by brainchild0** [comment]
@dzmitryv: Are these detailed results different from what you would expect?

**Comment by dzmitryv** [comment]
@brainchild0, yes, those results are different from what I would expect. I would expect all those 9 links to work as they do in Atom for instance or here on GitHub:

1. inline link [foo](http://go/test) 
2. bare URI http://go/test  
3. autolink <http://go/test>
4. [gitHub](https://github.com)
5. https://github.com
6. <https://github.com>
7. [gitHub](http://github.com)
8. http://github.com
9. <http://github.com>

**Comment by brainchild0** [comment]
@dzmitryv: It was suggested that the difference is likely to be, at least in part, related to some of the  items in your example not being valid URIs. Perhaps you might explain in more detail, if your view is that the current behavior is inadequate.

**Comment by dzmitryv** [comment]
@brainchild0 
In my opinion, the current behavior is inadequate because:
1. the editor's behavior is inconsistent - even among items 4-9 in my examples (I didn't see a comment that would clarify which of the URLs I've shown are not valid... I hope not those), pop-up icon appears only for 2 our of 6 items. I guess, I'm just reiterating [comment above](https://github.com/marktext/marktext/issues/2165#issuecomment-647244853).
1. the current behavior makes MarkText a bit too limited for a group of users who would expect links that work in other software (Chrome, GitHub, Markdown Preview Enhanced package in Atom) to also work in MarkText. I guess, I'm talking about non-working examples 1-3. I mean, from a user perspective, it's expected that something that he is using on daily basis in one or two reasonably popular applications like Chrome and GitHub would also work in other applications.

**Comment by brainchild0** [comment]
@dzmitryv: The item `http://go/test` is not a valid URI, as `go` is not a valid top-level domain name. Such is the reason why it is treated differently in MarkText than other items in your list. Especially in the case when the item is given inline, without any syntax to indicate a link, it is treated as plain text. Such behavior may be desirable, even if different from other applications.

I believe there is a case for more consistent behavior in MarkText than currently, but it may help for you to be precise about what you expect in each case. It may be desirable not to treat items the same regardless of whether they are invalid URIs, such as in the case of `//` not being followed by a full domain name.

**Comment by alaestor** [comment]
Those two buttons have been a frustrating UX problem for me. I frequently have lists of links and I **_constantly_** end up miss-clicking on the popup button and breaking or opening the link above the one i intended to move the cursor to or select. The popup seems to be aware of my frustration and is a bit temperamental about it, so it decides to vanish rapidly whenever I actually want to use the darn thing.

Currently I've began to work around this by always click _past_ the end of the link when I intend to edit or select it, and ctrl+clicking _on_ the link when I wish to navigate; but it's still a clunky annoyance.

Just some ideas:

1. right-click context menu, ctrl-click on to navigate, click to edit
2. hold a key and hover for the pop-up, ctrl-click on to navigate, click to edit
3. no pop-up, click on to navigate, click past to edit
 
The first option would be preferable. The third option is the simplest to implement and easiest to maintain because it's very close to the current behavior and it doesn't rely on modifier key detection which can have portability complications; it just gets rid of the annoying behavior but has the downside of requiring user discretion when clicking around the glowing bits of text (which could be viewed as a security issue...)

**Comment by amterp** [comment]
I'd like to voice my support for links like `http://go/test` to be clickable same as `https://github.com/` is. My company has a lot of internal servers that provide links which MarkText doesn't seem to like for whatever reason, but Obsidian *would* make clickable, as desired.
